pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -5489,4 +5489,114 @@ protected CompletableFuture<AutoSubscriptionCreationOverride> internalGetAutoSub
                         }));
     }
+
+    protected CompletableFuture<MessageId> internalGetMessageIDByIndexAsync(Long index, boolean authoritative) {
+        if (!pulsar().getBrokerService().isBrokerEntryMetadataEnabled()) {
+            return FutureUtil.failedFuture(new RestException(Status.PRECONDITION_FAILED,
+                    "GetMessageIDByIndex is not allowed when broker entry metadata is disabled"));
+        }
+        if (index == null || index < 0) {
+            return FutureUtil.failedFuture(new RestException(Status.NOT_FOUND,
+                    "Invalid message index: " + index));
+        }
+        int partitionIndex = topicName.getPartitionIndex();
+        CompletableFuture<Void> future = validateTopicOperationAsync(topicName, TopicOperation.PEEK_MESSAGES);
+        return future.thenCompose(__ -> {
+                    if (topicName.isGlobal()) {
+                        return validateGlobalNamespaceOwnershipAsync(namespaceName);
+                    } else {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                }).thenCompose(__ -> {
+                    if (topicName.isPartitioned()) {
+                        return CompletableFuture.completedFuture(null);
+                    } else {
+                        return getPartitionedTopicMetadataAsync(topicName, authoritative, false)
+                                .thenAccept(topicMetadata -> {
+                                    if (topicMetadata.partitions > 0) {
+                                        log.warn("[{}] Not supported getMessageIdByIndex operation on "
+                                                        + "partitioned-topic {}", clientAppId(), topicName);
+                                        throw new RestException(Status.METHOD_NOT_ALLOWED,
+                                                "GetMessageIDByIndex is not allowed on partitioned-topic");
+                                    }
+                                });
+                    }
+                }).thenCompose(ignore -> validateTopicOwnershipAsync(topicName, authoritative))
+                .thenCompose(__ -> getTopicReferenceAsync(topicName))
+                .thenCompose(topic -> {
+                    if (!(topic instanceof PersistentTopic persistentTopic)) {
+                        log.error("[{}] Get message id by index on a non-persistent topic {} is not allowed",
+                                clientAppId(), topicName);
+                        return FutureUtil.failedFuture(new RestException(Status.METHOD_NOT_ALLOWED,
+                                "Get message id by index on a non-persistent topic is not allowed"));
+                    }
+                    ManagedLedger managedLedger = persistentTopic.getManagedLedger();
+                    return findMessageIndexByPosition(
+                            PositionFactory.create(managedLedger.getFirstPosition().getLedgerId(), 0),
```

Comment:
Why not just pass `managedLedger.getFirstPosition()` here?

Reply:
IMO, the method `findMessageIndexByPosition` may be reused in the future,  so it need a param `position` instead of only managerLedger

Reply:
Sorry I've thought that `managedLedger.getFirstPosition()` is the same with `PositionFactory.create(managedLedger.getFirstPosition().getLedgerId(), 0)`, so I asked the question.

But now there is a new question, what if the ledger of the first position is empty (i.e. topic is empty)? In this case, `asyncReadEntry` could fail because the entry does not exist. Could you add a test to show the behavior when `getMessageIdByIndex` is called on an empty topic?

Reply:
`findMessageIndexByPosition(managedLedger.getFirstPosition(), ...)` is better.

Reply:
>Could you add a test to show the behavior when getMessageIdByIndex is called on an empty topic?

Done

Reply:
>findMessageIndexByPosition(managedLedger.getFirstPosition(), ...) is better.

@nodece `managedLedger.getFirstPosition()` will return [first ledger Id, -1]